### DISCUSSION TITLE
Shortest paths: Add running cost to the signature of a weight function

### DISCUSSION THREAD

**Discussion by harisbal** [networkx/networkx#7080] [Ideas]
Hello,
I am facing a situation where the cost of a path should be reset at specific nodes (_charging nodes_).

I was thinking that it would be easy to reset the cost of paths during shortest path identification if the `weight` function was also provided with the running cost of the path. If so, the weight of the edge destinating to a charging node could be set to zero using the function below:

```python
def _weight(u, v, d, c):
    # v is a charging node
    if d['type'] == 'charge_link':
        return -c + d['weight']
    else:
        return d['weight']
```

A simple modification below (adding the `c` parameter), could work for non user-defined functions
https://github.com/networkx/networkx/blob/6ced50a517f4724d774c4021f45b961ba639c387/networkx/algorithms/shortest_paths/weighted.py#L76-L78

changes to:

```python
if G.is_multigraph():
    return lambda u, v, d, c: min(attr.get(weight, 1) for attr in d.values())
return lambda u, v, data, c: data.get(weight, 1)
```

For the case of dijkstra, we just need to change
https://github.com/networkx/networkx/blob/6ced50a517f4724d774c4021f45b961ba639c387/networkx/algorithms/shortest_paths/weighted.py#L851-L855

to

```python
dist[v] = d
if v == target:
    break
for u, e in G_succ[v].items():
    cost = weight(v, u, e, d)
```

I am quite uncertain about how we should treat the user-defined weight functions so I leave this open to discussion.
Thoughts?

Cheers

**Comment by dschult** [comment]
I think there are some unsolvable issues about the stated problem you are trying to solve: When the cost gets set to zero at a node, that means that going back along an already traveled edge will lead to a much different cost than it used to. In a fundamental sense, you are no longer looking at a shortest_path. Like if I have to go through Albany to get to the charging station in Troy, and then I go back to albany, which "distance" am I minimizing? the distance from the original node? Or the distance from Troy? Somehow you need to track both the distance and how many times you had to charge.  Perhaps this is a collection of many shortest_path problems from each of the charging stations. 

But... that said, the weight function system is quite flexible... especially if you use some trickery with variables that point to objects which are updated. For example, before calling dijkstra, create a dict to hold the costs-so-far: `cost_so_far=collections.defaultdict(int)`. Then write your `weight` function to compute the weight as it usually does, but then before returning that weight, update the dict: `cost_so_far[v] = min(cost_so_far[u] + wt, cost_so_far[v])` (or something similar). Notice that `cost_so_far` is not an input to `weight`. But it is still available to `weight` if it is defined in the function/module that your weight function is defined in. By manually updating the cost_so_far each time you look at an edge to that node, you will get a beter and better idea of what the shortest cost is to that node. I believe with Dijkstra, you will get the shortest distance on the first call so the code might be shorter and simpler.

NetworkX does not provide a way for you to provide data other than u, v, and edgedata to the weight function. But it is possible for that function to use information not provided by the calling environment.  And I believe you need to think hard about what you mean by shortest_path when you talk about resetting the distance when the search reaches a specific node.

**Reply by harisbal** [reply]
Hello @dschult ,
Thank you for this elaborate answer. You made me realise that 

1. I had oversimplified the concept
2. We can keep track of the total cost to the node and dynamically set edge costs without interferring with the algorithm.
3. the actual problem I'm interested in is the enumeration of all possible paths which do not violate a capacity constraint. Mentioning `shortest_path` derailed the disscussion.

I understand that including the cost so far in the weight function signature is not required so I close the discussion for now.
